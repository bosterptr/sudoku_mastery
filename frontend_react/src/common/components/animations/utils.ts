import { CSSProperties } from 'react';

export const reflow = (node: Element) => node.scrollTop;

interface ComponentProps {
  easing: string | { enter?: string; exit?: string } | undefined;
  style: CSSProperties | undefined;
  timeout: number | { enter?: number; exit?: number };
}

interface Options {
  mode: 'enter' | 'exit';
}

interface TransitionProps {
  duration: string | number;
  easing: string | undefined;
  delay: string | undefined;
}

export function getTransitionProps(props: ComponentProps, options: Options): TransitionProps {
  const { timeout, easing, style = {} } = props;

  return {
    duration:
      style.transitionDuration ??
      (typeof timeout === 'number' ? timeout : timeout[options.mode] || 0),
    easing:
      style.transitionTimingFunction ??
      (typeof easing === 'object' ? easing[options.mode] : easing),
    delay: style.transitionDelay,
  };
}

export function getAutoHeightDuration(height: number) {
  if (!height) return 0;
  const constant = height / 36;
  // https://www.wolframalpha.com/input/?i=(4+%2B+15+*+(x+%2F+36+)+**+0.25+%2B+(x+%2F+36)+%2F+5)+*+10
  return Math.round((4 + 15 * constant ** 0.25 + constant / 5) * 10);
}
