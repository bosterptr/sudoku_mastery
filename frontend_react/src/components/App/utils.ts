const GL_BUFFER_STATUS_CHECK_INTERVAL = 2;

export type IOpenglSetup = [
  OffscreenCanvas,
  WebGLProgram,
  WebGLProgram,
  WebGLProgram,
  ITexFbPair,
  ITexFbPair,
  number,
];

export interface ITexFbPair {
  tex: WebGLTexture;
  fb: WebGLFramebuffer;
}
export const getGrayScaleFromUint8Array = (
  dataArray: Uint8Array,
  width: number,
  height: number,
  cachedGrayScaleUintArray: Uint8Array,
  cachedGrayScaleSize: number,
): Uint8Array => {
  if (cachedGrayScaleSize !== width * height) {
    // eslint-disable-next-line no-param-reassign
    cachedGrayScaleUintArray = new Uint8Array(width * height);
    // eslint-disable-next-line no-param-reassign
    cachedGrayScaleSize = width * height;
  }
  const { length } = dataArray;
  for (let i = 0, luminanceIndex = 0; i < length; i += 4, luminanceIndex += 1) {
    // eslint-disable-next-line no-bitwise
    const r = dataArray[i] >> 2;
    // eslint-disable-next-line no-bitwise
    const g = dataArray[i + 1] >> 1;
    // eslint-disable-next-line no-bitwise
    const b = dataArray[i + 2] >> 3;
    // const luminance = 0.299 * r + 0.587 * g + 0.114 * b;
    // luminanceData[luminanceIndex] =
    //   0.299 * dataArray[i] +
    //   0.587 * dataArray[i + 1] +
    //   0.114 * dataArray[i + 2];
    // eslint-disable-next-line no-param-reassign
    cachedGrayScaleUintArray[luminanceIndex] = r + b + g;
  }
  return cachedGrayScaleUintArray;
};

export const clientWaitAsync = (
  gl: WebGL2RenderingContext,
  sync: WebGLSync,
  flags = 0,
  intervalMs = 10,
) => {
  return new Promise<void>((resolve, reject) => {
    const check = () => {
      const res = gl.clientWaitSync(sync, flags, 0);
      if (res === gl.WAIT_FAILED) {
        reject();
        return;
      }
      if (res === gl.TIMEOUT_EXPIRED) {
        setTimeout(check, intervalMs);
        return;
      }
      resolve();
    };
    check();
  });
};

export const readPixelsAsync = (
  gl: WebGL2RenderingContext,
  width: number,
  height: number,
  buffer: Uint8Array,
): Promise<Uint8Array> => {
  const bufpak = gl.createBuffer();
  gl.bindBuffer(gl.PIXEL_PACK_BUFFER, bufpak);
  gl.bufferData(gl.PIXEL_PACK_BUFFER, buffer.byteLength, gl.STREAM_READ);
  gl.readPixels(0, 0, width, height, gl.RGBA, gl.UNSIGNED_BYTE, 0);
  const sync = gl.fenceSync(gl.SYNC_GPU_COMMANDS_COMPLETE, 0);
  if (!sync) throw new Error('WebGLSync is empty');
  gl.flush();
  return clientWaitAsync(gl, sync, 0, GL_BUFFER_STATUS_CHECK_INTERVAL).then(() => {
    gl.deleteSync(sync);
    gl.bindBuffer(gl.PIXEL_PACK_BUFFER, bufpak);
    gl.getBufferSubData(gl.PIXEL_PACK_BUFFER, 0, buffer);
    gl.bindBuffer(gl.PIXEL_PACK_BUFFER, null);
    gl.deleteBuffer(bufpak);
    return buffer;
  });
};

export function setAttributes(gl: WebGL2RenderingContext, program: WebGLProgram, name: string) {
  const location = gl.getAttribLocation(program, name);
  gl.enableVertexAttribArray(location);
  gl.vertexAttribPointer(location, 2, gl.FLOAT, false, 0, 0);
}

export function isMobile() {
  const userAgent = navigator.userAgent.toLowerCase();
  return /iphone|ipod|android|ie|blackberry|fennec/.test(userAgent);
}
export const getCameraResolution = () => {
  if (isMobile()) {
    return {
      width: { ideal: document.body.clientWidth > 500 ? 500 : document.body.clientWidth },
      height: { ideal: document.body.clientHeight > 500 ? 500 : document.body.clientHeight },
    };
  }
  return {
    width: { ideal: 1000 },
    height: { ideal: 1000 },
  };
};
