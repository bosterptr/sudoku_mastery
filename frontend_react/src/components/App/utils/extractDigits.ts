import { LayersModel, loadLayersModel, reshape, tensor2d } from '@tensorflow/tfjs';

const indexOfMaxValue = <T>(
  a: T[],
  func: (param: T) => number = ((i) => i) as (param: T) => number,
) => a.reduce((iMax, x, i, arr) => (func(x) > func(arr[iMax]) ? i : iMax), 0);

let model: LayersModel;
const CELLS_IN_ROW = 9;
const CELL_WIDTH = 20;
const TOTAL_CELLS = CELLS_IN_ROW * CELLS_IN_ROW;
const IMAGE_WIDTH = CELLS_IN_ROW * CELL_WIDTH;
const CELL_PIXEL_COUNT = CELL_WIDTH * CELL_WIDTH;

function extractCube(src: Uint8Array, x: number, y: number): Uint8Array {
  if (x < 0 || y < 0 || x + CELL_WIDTH > IMAGE_WIDTH || y + CELL_WIDTH > IMAGE_WIDTH) {
    throw new Error('Starting point out of bounds');
  }
  const cube = new Uint8Array(CELL_WIDTH * CELL_WIDTH);
  for (let row = 0; row < CELL_WIDTH; row += 1) {
    for (let col = 0; col < CELL_WIDTH; col += 1) {
      const srcIndex = (y + row) * IMAGE_WIDTH + (x + col);
      const cubeIndex = row * CELL_WIDTH + col;
      cube[cubeIndex] = src[srcIndex];
    }
  }

  return cube;
}

const init = async () => {
  model = await loadLayersModel('./models/model6.json');
};
const extractDigits = (cells: Uint8Array) => {
  const testDataArray = new Float32Array(IMAGE_WIDTH * IMAGE_WIDTH);
  for (let i = 0; i < TOTAL_CELLS; i += 1) {
    const x = (i % CELLS_IN_ROW) * CELL_WIDTH;
    const y = Math.floor(i / CELLS_IN_ROW) * CELL_WIDTH;
    const buffer = extractCube(cells, x, y);
    testDataArray.set(buffer, i * CELL_PIXEL_COUNT);
  }
  const testTensor = tensor2d(testDataArray, [TOTAL_CELLS, CELL_PIXEL_COUNT]);
  const reshaped = reshape(testTensor, [TOTAL_CELLS, CELL_WIDTH, CELL_WIDTH, 1]);
  const tensor = model.predict(reshaped);
  if (Array.isArray(tensor)) throw new Error();
  const prediction = tensor.dataSync();
  const predictionData: number[] = [];
  for (const num of prediction) {
    predictionData.push(num);
  }
  let result = '';
  for (let i = 0; i < TOTAL_CELLS; i += 1) {
    const cellPrediction = predictionData.slice(i * 10, i * 10 + 10);
    const digit = indexOfMaxValue(cellPrediction);
    result += digit === 0 ? '.' : digit;
  }
  return result;
};

export { extractDigits, init };
