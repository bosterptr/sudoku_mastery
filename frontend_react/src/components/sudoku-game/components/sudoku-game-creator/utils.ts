import { ALL_SUDOKU_CELLS, CANVAS_SIZE, SUDOKU_CELLS_IN_ROW } from 'app/core/constants/sudoku';
import { ICellNotes, IMission } from './types';

export const generateNotes = (): ICellNotes[] => {
  const newArray: ICellNotes[] = [];
  for (let cellIndex = 0; cellIndex < ALL_SUDOKU_CELLS; cellIndex += 1) {
    newArray.push([]);
  }
  return newArray;
};
export const getBlockSize = (pixelRatio: number) => (CANVAS_SIZE * pixelRatio - 2 * pixelRatio) / 3;
export const getAdjustedPixelRatio = () => {
  const { devicePixelRatio } = window;
  return devicePixelRatio > 2 ? 2 * devicePixelRatio : 4;
};
export const getCellSize = (pixelRatio: number) => (CANVAS_SIZE * pixelRatio - 4 * pixelRatio) / 9;
export const getContextSize = (pixelRatio: number) => CANVAS_SIZE * pixelRatio;
export const getOuterSize = (pixelRatio: number) => 2 * pixelRatio;
export const getThinSize = (pixelRatio: number) => 0.75 * pixelRatio;
export const getX = (index: number) => Math.floor(index / 9);
export const getY = (index: number) => index % 9;
export const getCellXPosition = (outerSize: number, cellSize: number, index: number) =>
  outerSize + cellSize * getY(index);
export const getCellYPosition = (outerSize: number, cellSize: number, index: number) =>
  outerSize + cellSize * getX(index);
export const getCubeId = (index: number) =>
  `${Math.floor(getX(index) / 3)}${Math.floor(getY(index) / 3)}`;
export const EMPTY_CELL = '.';
export const isSudokuCellEmpty = (char: string) => char === EMPTY_CELL;
export const generateArrWithTrueAtIndexes = (indexes: number[]): boolean[] => {
  const newArray: boolean[] = [];
  for (let cellIndex = 0; cellIndex < ALL_SUDOKU_CELLS; cellIndex += 1) {
    if (indexes.includes(cellIndex)) newArray[cellIndex] = true;
    else newArray[cellIndex] = false;
  }
  return newArray;
};
export const getScoreFromMove = (values: IMission, solution: string, selectedCell: number) => {
  let fullRows = 0;
  let fullColumns = 0;
  let fullCubes = 0;
  for (let cellIndex = 0; cellIndex < ALL_SUDOKU_CELLS; cellIndex += 1) {
    if (values[cellIndex] === solution[cellIndex]) {
      if (getY(cellIndex) === getY(selectedCell)) fullRows += 1;
      if (getX(cellIndex) === getX(selectedCell)) fullColumns += 1;
      if (getCubeId(cellIndex) === getCubeId(selectedCell)) fullCubes += 1;
    }
  }
  let score = 1;
  if (fullRows === SUDOKU_CELLS_IN_ROW) score += 5;
  if (fullColumns === SUDOKU_CELLS_IN_ROW) score += 5;
  if (fullCubes === 9) score += 5;
  return score;
};
