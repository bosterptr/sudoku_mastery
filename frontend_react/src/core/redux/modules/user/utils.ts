import { IUser } from 'app/core/redux/modules/user/types';

// eslint-disable-next-line import/prefer-default-export
export function normalizeSetUsers(users: Record<string, IUser>) {
  const normalizedUsers: Record<string, IUser> = {};
  Object.entries(users).forEach(([, user]) => {
    normalizedUsers[user.id] = user;
  });

  return normalizedUsers;
}
