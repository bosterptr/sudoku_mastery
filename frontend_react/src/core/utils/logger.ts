export interface Logger {
  withCategory(category: string): Logger;
  withComponentName(componentName: string): Logger;
  debug(message: string, ...args: unknown[]): void;
  info(message: string, ...args: unknown[]): void;
  info(message: string, ...args: unknown[]): void;
  warn(message: string, ...args: unknown[]): void;
  errorAndReport(error: Readonly<Error>, message: string, ...args: unknown[]): void;
}
