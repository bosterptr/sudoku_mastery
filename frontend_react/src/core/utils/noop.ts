export default function NOOP(): void {}
