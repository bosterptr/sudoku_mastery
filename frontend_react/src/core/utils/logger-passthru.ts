import { Logger } from './logger';

/* eslint-disable no-console */
export class PassthruLogger {
  public withCategory(): Logger {
    return this;
  }

  public withComponentName(): Logger {
    return this;
  }

  // eslint-disable-next-line class-methods-use-this
  public debug(message?: unknown) {
    console.debug(message);
  }

  // eslint-disable-next-line class-methods-use-this
  public info(message?: unknown) {
    console.log(message);
  }

  // eslint-disable-next-line class-methods-use-this
  public warn(message?: unknown) {
    console.warn(message);
  }

  // eslint-disable-next-line class-methods-use-this
  public errorAndReport(error: Error, message: string) {
    // console.error(error, message);
    // console.log(error, message);
    // eslint-disable-next-line no-unused-expressions
    error.message === message;
  }
}
