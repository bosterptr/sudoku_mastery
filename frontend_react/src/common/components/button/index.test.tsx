import { fireEvent, render, screen } from 'app/tests/utils';
import '@testing-library/jest-dom';
import Button from '.';

describe('Button', () => {
  it('renders button with children', () => {
    render(<Button>Click me</Button>);
    expect(screen.getByRole('button')).toHaveTextContent('Click me');
    expect(screen.getByRole('button')).toMatchSnapshot();
  });

  it('calls onClick prop when clicked', () => {
    const handleClick = jest.fn();
    render(<Button onClick={handleClick}>Click me</Button>);
    fireEvent.click(screen.getByText('Click me'));
    expect(handleClick).toHaveBeenCalledTimes(1);
  });
  test('renders variants and categories correctly', () => {
    const { rerender } = render(
      <Button variant="success" category="primary">
        Success
      </Button>,
    );
    expect(screen.getByText('Success')).toHaveStyle('background-color: transparent');

    rerender(
      <Button variant="danger" category="secondary">
        Danger
      </Button>,
    );
    expect(screen.getByText('Danger')).toHaveStyle('color: rgb(255, 255, 255)');

    rerender(
      <Button variant="warning" category="secondary">
        Warning
      </Button>,
    );
    expect(screen.getByText('Warning')).toHaveStyle('color: rgb(255, 255, 255)');

    rerender(
      <Button variant="info" category="secondary">
        Info
      </Button>,
    );
    expect(screen.getByText('Info')).toHaveStyle('color: rgb(255, 255, 255)');
  });
  it('is disabled when disabled prop is true', () => {
    const { getByRole } = render(<Button disabled>Disabled Button</Button>);
    expect(getByRole('button')).toBeDisabled();
  });
  it('shows spinner when isLoading is true', () => {
    const { queryByText, getByTestId } = render(<Button isLoading>Loading...</Button>);
    expect(queryByText('Loading...')).not.toBeInTheDocument();
    expect(getByTestId('spinner')).toBeInTheDocument();
  });
  test('renders sizes correctly', () => {
    const { rerender } = render(<Button size="lg">lg</Button>);
    expect(screen.getByText('lg')).toHaveStyle('padding: 10px 13px');
    expect(screen.getByText('lg')).toMatchSnapshot();
    rerender(<Button size="md">md</Button>);
    expect(screen.getByText('md')).toHaveStyle('padding: 6px 8px');
    expect(screen.getByText('md')).toMatchSnapshot();
    rerender(<Button size="sm">sm</Button>);
    expect(screen.getByText('sm')).toHaveStyle('padding: 4px 6px');
    expect(screen.getByText('sm')).toMatchSnapshot();
    rerender(<Button size="xs">xs</Button>);
    expect(screen.getByText('xs')).toHaveStyle('padding: 2px 4px');
    expect(screen.getByText('xs')).toMatchSnapshot();
  });
  test('renders icons', () => {
    expect(render(<Button iconAfter={<div>Icon</div>} />).asFragment());
    expect(screen.getByRole('button')).toMatchSnapshot();
    expect(screen.getByText('Icon')).toBeDefined();
  });
});
