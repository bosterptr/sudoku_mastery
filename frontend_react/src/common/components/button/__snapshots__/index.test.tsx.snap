// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders button with children 1`] = `
.c0 {
  align-items: center;
  border-radius: 3px;
  border: none;
  box-shadow: none;
  box-sizing: border-box;
  cursor: pointer;
  display: inline-flex;
  font-weight: 500;
  justify-content: center;
  line-height: 1.75;
  text-decoration: none;
  transition: color 0.2s ease,background-color 0.2s ease,box-shadow 0.2s ease,border-color 0.2s ease,filter 0.2s ease;
  user-select: none;
  background-color: #1648ff;
  color: #fff;
  fill: #1648ff;
  border: 2px solid transparent;
  transition: background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  font-size: 0.875rem;
  padding: 6px 8px;
}

.c0:focus:not {
  outline: 0;
}

.c0:hover {
  background-color: #1236bb;
  color: #fff;
}

.c0:active {
  background-color: transparent;
  color: #1648ff;
  box-shadow: 0px 4px 8px #181a1b33;
}

.c0>* {
  stroke: #1648ff;
  fill: #fff;
}

<button
  class="c0"
  color="rgba(0, 0, 0, .2)"
  type="button"
>
  Click me
</button>
`;

exports[`Button renders icons 1`] = `
.c0 {
  align-items: center;
  border-radius: 3px;
  border: none;
  box-shadow: none;
  box-sizing: border-box;
  cursor: pointer;
  display: inline-flex;
  font-weight: 500;
  justify-content: center;
  line-height: 1.75;
  text-decoration: none;
  transition: color 0.2s ease,background-color 0.2s ease,box-shadow 0.2s ease,border-color 0.2s ease,filter 0.2s ease;
  user-select: none;
  background-color: #1648ff;
  color: #fff;
  fill: #1648ff;
  border: 2px solid transparent;
  transition: background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  font-size: 0.875rem;
  padding: 6px 8px;
}

.c0:focus:not {
  outline: 0;
}

.c0:hover {
  background-color: #1236bb;
  color: #fff;
}

.c0:active {
  background-color: transparent;
  color: #1648ff;
  box-shadow: 0px 4px 8px #181a1b33;
}

.c0>* {
  stroke: #1648ff;
  fill: #fff;
}

<button
  class="c0"
  color="rgba(0, 0, 0, .2)"
  type="button"
>
  <div>
    Icon
  </div>
</button>
`;

exports[`Button renders sizes correctly 1`] = `
.c0 {
  align-items: center;
  border-radius: 3px;
  border: none;
  box-shadow: none;
  box-sizing: border-box;
  cursor: pointer;
  display: inline-flex;
  font-weight: 500;
  justify-content: center;
  line-height: 1.75;
  text-decoration: none;
  transition: color 0.2s ease,background-color 0.2s ease,box-shadow 0.2s ease,border-color 0.2s ease,filter 0.2s ease;
  user-select: none;
  background-color: #1648ff;
  color: #fff;
  fill: #1648ff;
  border: 2px solid transparent;
  transition: background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  font-size: 0.9375rem;
  padding: 10px 13px;
}

.c0:focus:not {
  outline: 0;
}

.c0:hover {
  background-color: #1236bb;
  color: #fff;
}

.c0:active {
  background-color: transparent;
  color: #1648ff;
  box-shadow: 0px 4px 8px #181a1b33;
}

.c0>* {
  stroke: #1648ff;
  fill: #fff;
}

<button
  class="c0"
  color="rgba(0, 0, 0, .2)"
  type="button"
>
  lg
</button>
`;

exports[`Button renders sizes correctly 2`] = `
.c0 {
  align-items: center;
  border-radius: 3px;
  border: none;
  box-shadow: none;
  box-sizing: border-box;
  cursor: pointer;
  display: inline-flex;
  font-weight: 500;
  justify-content: center;
  line-height: 1.75;
  text-decoration: none;
  transition: color 0.2s ease,background-color 0.2s ease,box-shadow 0.2s ease,border-color 0.2s ease,filter 0.2s ease;
  user-select: none;
  background-color: #1648ff;
  color: #fff;
  fill: #1648ff;
  border: 2px solid transparent;
  transition: background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  font-size: 0.875rem;
  padding: 6px 8px;
}

.c0:focus:not {
  outline: 0;
}

.c0:hover {
  background-color: #1236bb;
  color: #fff;
}

.c0:active {
  background-color: transparent;
  color: #1648ff;
  box-shadow: 0px 4px 8px #181a1b33;
}

.c0>* {
  stroke: #1648ff;
  fill: #fff;
}

<button
  class="c0"
  color="rgba(0, 0, 0, .2)"
  type="button"
>
  md
</button>
`;

exports[`Button renders sizes correctly 3`] = `
.c0 {
  align-items: center;
  border-radius: 3px;
  border: none;
  box-shadow: none;
  box-sizing: border-box;
  cursor: pointer;
  display: inline-flex;
  font-weight: 500;
  justify-content: center;
  line-height: 1.75;
  text-decoration: none;
  transition: color 0.2s ease,background-color 0.2s ease,box-shadow 0.2s ease,border-color 0.2s ease,filter 0.2s ease;
  user-select: none;
  background-color: #1648ff;
  color: #fff;
  fill: #1648ff;
  border: 2px solid transparent;
  transition: background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  font-size: 0.8125rem;
  padding: 4px 6px;
}

.c0:focus:not {
  outline: 0;
}

.c0:hover {
  background-color: #1236bb;
  color: #fff;
}

.c0:active {
  background-color: transparent;
  color: #1648ff;
  box-shadow: 0px 4px 8px #181a1b33;
}

.c0>* {
  stroke: #1648ff;
  fill: #fff;
}

<button
  class="c0"
  color="rgba(0, 0, 0, .2)"
  type="button"
>
  sm
</button>
`;

exports[`Button renders sizes correctly 4`] = `
.c0 {
  align-items: center;
  border-radius: 3px;
  border: none;
  box-shadow: none;
  box-sizing: border-box;
  cursor: pointer;
  display: inline-flex;
  font-weight: 500;
  justify-content: center;
  line-height: 1.75;
  text-decoration: none;
  transition: color 0.2s ease,background-color 0.2s ease,box-shadow 0.2s ease,border-color 0.2s ease,filter 0.2s ease;
  user-select: none;
  background-color: #1648ff;
  color: #fff;
  fill: #1648ff;
  border: 2px solid transparent;
  transition: background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  font-size: 0.6875rem;
  padding: 2px 4px;
}

.c0:focus:not {
  outline: 0;
}

.c0:hover {
  background-color: #1236bb;
  color: #fff;
}

.c0:active {
  background-color: transparent;
  color: #1648ff;
  box-shadow: 0px 4px 8px #181a1b33;
}

.c0>* {
  stroke: #1648ff;
  fill: #fff;
}

<button
  class="c0"
  color="rgba(0, 0, 0, .2)"
  type="button"
>
  xs
</button>
`;
